modules/rostests/kmtests/ntos_se/SeLogonSession.c

```diff
@@ -0,0 +1,35 @@
+/*
+ * PROJECT:     ReactOS kernel-mode tests
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Kernel mode tests for logon session manipulation API
+ * COPYRIGHT:   Copyright 2021 George Bișoc <george.bisoc@reactos.org>
+ */
+
+#include <kmt_test.h>
+#include <ntifs.h>
+
+#define IMAGINARY_LOGON {0x001, 0}
+
+static
+VOID
+LogonMarkTermination(VOID)
+{
+    NTSTATUS Status;
+    LUID SystemLogonID = SYSTEM_LUID;
+    LUID NonExistentLogonID = IMAGINARY_LOGON;
+
+    /* Mark the system authentication logon for termination */
+    Status = SeMarkLogonSessionForTerminationNotification(&SystemLogonID);
+    ok_irql(PASSIVE_LEVEL);
+    ok_eq_hex(Status, STATUS_SUCCESS);
+
+    /* We give a non existent logon */
+    Status = SeMarkLogonSessionForTerminationNotification(&NonExistentLogonID);
```

Comment:
I'm wondering, whether it is possible to artificially open, then close a small logon session, in order to verify whether the notification will be raised.

Reply:
I doubt one can simply artificially open/create a logon session, since the internal part of the system is up to create sessions (and the LSA auth package for that matter but that is in user mode though) and not anybody else. But even then, session termination notification is far from being fully complete in ReactOS.

When a session is marked for termination, Windows notifies both LSA and filesystems that a session is about to terminate in logon dereference routine (that is `SepRmDereferenceLogonSession` in our case). This is done by using several worker queuing kernel routines like `ExQueueWorkItem` and such, which the kernel gives the opportunity to both LSA and FSDs to do whatever stuff they have to do once a logon is about to terminate. Our `SepRmDereferenceLogonSession` is not fully implemented because of that.

Given how this stuff works in Windows, my plan is to have it implemented accordingly in ReactOS thorough different PRs for the sake of easier review. First implement this `SeMarkLogonSessionForTerminationNotification` and then fully implement the logon dereference function which notifies LSA and FSDs thorough respective internal subroutines that deal with this stuff.

Reply:
You may consider writing to Eric in order to obtain advice on how creating/deleting logon sessions as tests can be achieved.

Reply:
Alternatively I'm wondering whether this is possible from kernel-mode.

Reply:
I talked with Eric and no, there's no other alternative possible to simulate a logon creation because of how logon implementation is done in Windows.